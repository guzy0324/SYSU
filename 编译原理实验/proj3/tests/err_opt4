/* Optional test Case 4: undeclared class methods f0() and f1() */
program erropt4;
class Point0 {
	 declarations
		int[] parr0 = int [12];
        	int x = 5, y;
	 enddeclarations
	 method void f0()
	 {
		system.println('Before x=');
		system.println(x); 
	        x := x*x;   
		y := 1;
		while (y<12)
		{
			parr0[y] := parr0[y-1]+1;
			y := y+1;
		};
		system.println('After x=');
		system.println(x); 
	 }
}

class Point1 {
	declarations
	        int t1 = 12;
            	Point0 p1;
	enddeclarations
	 method void f1()
	 {
		system.println('Before t1=');	
		system.println(t1);
	        t1   := 1024; 
		p1.f0(); 
		system.println('After t1=');
		system.println(t1);
	 }
}

class Point2 {
	declarations
            	Point1 p21;
		Point0 p20;
		int xy;
	enddeclarations		
        method void main()	
	     declarations	
		        int xx = 22;	
			Point1 px;
	     enddeclarations
	{
	   Point0.f0();
           p21.f1(); 
	   px.p1.x := xx;
	   px.p1.f1(); 
	   p21.t1 := 133;
	   p21.f0();
	   system.println('p21.p1.x=');
	   system.println(p21.p1.x);
	   p20.f0();
	   xy := 0;
	   while (xy <12)
	   {
		system.println('p20.parr0:');
		system.println(p20.parr0[xy]);
		xy := xy+1;
	   };
	   Point1.f1();
	   system.println('p20.x=');
	   system.println(p20.x); 
        }
}



