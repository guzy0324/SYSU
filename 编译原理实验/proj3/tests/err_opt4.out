Semantic Error--line: 75, symbol f1: is an undeclared field name.
Semantic Error--line: 75, symbol f0: is an undeclared field name.
********************************Symbol Table************************************

          Name Nest-Level  Tree-Node Predefined        Kind      Type      Value     Offset  Dimension     Argnum

  1     system          0                   yes      class                                                       
  2    println          1                   yes  procedure                                                      1
  3     Point0          0                            class                                                       
  4      parr0          1                            array  type-tree                                1           
  5          y          1                         variable  type-tree                                            
  6          x          1                         variable  type-tree                                            
  7         f0          1  proc-tree             procedure                                                      0
  8     Point1          0                            class                                                       
  9         t1          1                         variable  type-tree                                            
 10         p1          1                         variable  type-tree                                            
 11         f1          1  proc-tree             procedure                                                      0
 12     Point2          0                            class                                                       
 13        p21          1                         variable  type-tree                                            
 14        p20          1                         variable  type-tree                                            
 15         xy          1                         variable  type-tree                                            
 16       main          1  proc-tree             procedure                                                      0
 17         xx          2                         variable  type-tree                                            
 18         px          2                         variable  type-tree                                            
************* SYNTAX TREE PRINTOUT ***********

  +-[DUMMYnode]
R-[ProgramOp]
  |   +-[STNode,12,"Point2"]
  | +-[ClassDefOp]
  | | |           +-[DUMMYnode]
  | | |         +-[CommaOp]
  | | |         | |   +-[DUMMYnode]
  | | |         | | +-[SelectOp]
  | | |         | | | | +-[DUMMYnode]
  | | |         | | | +-[FieldOp]
  | | |         | | |   +-[STNode,6,"x"]
  | | |         | +-[VarOp]
  | | |         |   +-[STNode,14,"p20"]
  | | |       +-[RoutineCallOp]
  | | |       | |   +-[DUMMYnode]
  | | |       | | +-[SelectOp]
  | | |       | | | | +-[DUMMYnode]
  | | |       | | | +-[FieldOp]
  | | |       | | |   +-[STNode,2,"println"]
  | | |       | +-[VarOp]
  | | |       |   +-[STNode,1,"system"]
  | | |     +-[StmtOp]
  | | |     | |     +-[DUMMYnode]
  | | |     | |   +-[CommaOp]
  | | |     | |   | +-[STRINGNode,149,"p20.x="]
  | | |     | | +-[RoutineCallOp]
  | | |     | | | |   +-[DUMMYnode]
  | | |     | | | | +-[SelectOp]
  | | |     | | | | | | +-[DUMMYnode]
  | | |     | | | | | +-[FieldOp]
  | | |     | | | | |   +-[STNode,2,"println"]
  | | |     | | | +-[VarOp]
  | | |     | | |   +-[STNode,1,"system"]
  | | |     | +-[StmtOp]
  | | |     |   |   +-[DUMMYnode]
  | | |     |   | +-[RoutineCallOp]
  | | |     |   | | |   +-[DUMMYnode]
  | | |     |   | | | +-[SelectOp]
  | | |     |   | | | | | +-[DUMMYnode]
  | | |     |   | | | | +-[FieldOp]
  | | |     |   | | | |   +-[STNode,11,"f1"]
  | | |     |   | | +-[VarOp]
  | | |     |   | |   +-[STNode,8,"Point1"]
  | | |     |   +-[StmtOp]
  | | |     |     |         +-[NUMNode,1]
  | | |     |     |       +-[AddOp]
  | | |     |     |       | | +-[DUMMYnode]
  | | |     |     |       | +-[VarOp]
  | | |     |     |       |   +-[STNode,15,"xy"]
  | | |     |     |     +-[AssignOp]
  | | |     |     |     | |   +-[DUMMYnode]
  | | |     |     |     | | +-[VarOp]
  | | |     |     |     | | | +-[STNode,15,"xy"]
  | | |     |     |     | +-[AssignOp]
  | | |     |     |     |   +-[DUMMYnode]
  | | |     |     |   +-[StmtOp]
  | | |     |     |   | |     +-[DUMMYnode]
  | | |     |     |   | |   +-[CommaOp]
  | | |     |     |   | |   | |     +-[DUMMYnode]
  | | |     |     |   | |   | |   +-[SelectOp]
  | | |     |     |   | |   | |   | | +-[DUMMYnode]
  | | |     |     |   | |   | |   | +-[IndexOp]
  | | |     |     |   | |   | |   |   | +-[DUMMYnode]
  | | |     |     |   | |   | |   |   +-[VarOp]
  | | |     |     |   | |   | |   |     +-[STNode,15,"xy"]
  | | |     |     |   | |   | | +-[SelectOp]
  | | |     |     |   | |   | | | | +-[DUMMYnode]
  | | |     |     |   | |   | | | +-[FieldOp]
  | | |     |     |   | |   | | |   +-[STNode,4,"parr0"]
  | | |     |     |   | |   | +-[VarOp]
  | | |     |     |   | |   |   +-[STNode,14,"p20"]
  | | |     |     |   | | +-[RoutineCallOp]
  | | |     |     |   | | | |   +-[DUMMYnode]
  | | |     |     |   | | | | +-[SelectOp]
  | | |     |     |   | | | | | | +-[DUMMYnode]
  | | |     |     |   | | | | | +-[FieldOp]
  | | |     |     |   | | | | |   +-[STNode,2,"println"]
  | | |     |     |   | | | +-[VarOp]
  | | |     |     |   | | |   +-[STNode,1,"system"]
  | | |     |     |   | +-[StmtOp]
  | | |     |     |   |   |     +-[DUMMYnode]
  | | |     |     |   |   |   +-[CommaOp]
  | | |     |     |   |   |   | +-[STRINGNode,138,"p20.parr0:"]
  | | |     |     |   |   | +-[RoutineCallOp]
  | | |     |     |   |   | | |   +-[DUMMYnode]
  | | |     |     |   |   | | | +-[SelectOp]
  | | |     |     |   |   | | | | | +-[DUMMYnode]
  | | |     |     |   |   | | | | +-[FieldOp]
  | | |     |     |   |   | | | |   +-[STNode,2,"println"]
  | | |     |     |   |   | | +-[VarOp]
  | | |     |     |   |   | |   +-[STNode,1,"system"]
  | | |     |     |   |   +-[StmtOp]
  | | |     |     |   |     +-[DUMMYnode]
  | | |     |     | +-[LoopOp]
  | | |     |     | | | +-[NUMNode,12]
  | | |     |     | | +-[LTOp]
  | | |     |     | |   | +-[DUMMYnode]
  | | |     |     | |   +-[VarOp]
  | | |     |     | |     +-[STNode,15,"xy"]
  | | |     |     +-[StmtOp]
  | | |     |       |   +-[NUMNode,0]
  | | |     |       | +-[AssignOp]
  | | |     |       | | |   +-[DUMMYnode]
  | | |     |       | | | +-[VarOp]
  | | |     |       | | | | +-[STNode,15,"xy"]
  | | |     |       | | +-[AssignOp]
  | | |     |       | |   +-[DUMMYnode]
  | | |     |       +-[StmtOp]
  | | |     |         |   +-[DUMMYnode]
  | | |     |         | +-[RoutineCallOp]
  | | |     |         | | |   +-[DUMMYnode]
  | | |     |         | | | +-[SelectOp]
  | | |     |         | | | | | +-[DUMMYnode]
  | | |     |         | | | | +-[FieldOp]
  | | |     |         | | | |   +-[STNode,7,"f0"]
  | | |     |         | | +-[VarOp]
  | | |     |         | |   +-[STNode,14,"p20"]
  | | |     |         +-[StmtOp]
  | | |     |           |     +-[DUMMYnode]
  | | |     |           |   +-[CommaOp]
  | | |     |           |   | |     +-[DUMMYnode]
  | | |     |           |   | |   +-[SelectOp]
  | | |     |           |   | |   | | +-[DUMMYnode]
  | | |     |           |   | |   | +-[FieldOp]
  | | |     |           |   | |   |   +-[STNode,6,"x"]
  | | |     |           |   | | +-[SelectOp]
  | | |     |           |   | | | | +-[DUMMYnode]
  | | |     |           |   | | | +-[FieldOp]
  | | |     |           |   | | |   +-[STNode,10,"p1"]
  | | |     |           |   | +-[VarOp]
  | | |     |           |   |   +-[STNode,13,"p21"]
  | | |     |           | +-[RoutineCallOp]
  | | |     |           | | |   +-[DUMMYnode]
  | | |     |           | | | +-[SelectOp]
  | | |     |           | | | | | +-[DUMMYnode]
  | | |     |           | | | | +-[FieldOp]
  | | |     |           | | | |   +-[STNode,2,"println"]
  | | |     |           | | +-[VarOp]
  | | |     |           | |   +-[STNode,1,"system"]
  | | |     |           +-[StmtOp]
  | | |     |             |     +-[DUMMYnode]
  | | |     |             |   +-[CommaOp]
  | | |     |             |   | +-[STRINGNode,128,"p21.p1.x="]
  | | |     |             | +-[RoutineCallOp]
  | | |     |             | | |   +-[DUMMYnode]
  | | |     |             | | | +-[SelectOp]
  | | |     |             | | | | | +-[DUMMYnode]
  | | |     |             | | | | +-[FieldOp]
  | | |     |             | | | |   +-[STNode,2,"println"]
  | | |     |             | | +-[VarOp]
  | | |     |             | |   +-[STNode,1,"system"]
  | | |     |             +-[StmtOp]
  | | |     |               |   +-[DUMMYnode]
  | | |     |               | +-[RoutineCallOp]
  | | |     |               | | |   +-[DUMMYnode]
  | | |     |               | | | +-[SelectOp]
  | | |     |               | | | | | +-[DUMMYnode]
  | | |     |               | | | | +-[FieldOp]
  | | |     |               | | | |   +-[IDNode,25,"f0"]
  | | |     |               | | +-[VarOp]
  | | |     |               | |   +-[STNode,13,"p21"]
  | | |     |               +-[StmtOp]
  | | |     |                 |   +-[NUMNode,133]
  | | |     |                 | +-[AssignOp]
  | | |     |                 | | |     +-[DUMMYnode]
  | | |     |                 | | |   +-[SelectOp]
  | | |     |                 | | |   | | +-[DUMMYnode]
  | | |     |                 | | |   | +-[FieldOp]
  | | |     |                 | | |   |   +-[STNode,9,"t1"]
  | | |     |                 | | | +-[VarOp]
  | | |     |                 | | | | +-[STNode,13,"p21"]
  | | |     |                 | | +-[AssignOp]
  | | |     |                 | |   +-[DUMMYnode]
  | | |     |                 +-[StmtOp]
  | | |     |                   |   +-[DUMMYnode]
  | | |     |                   | +-[RoutineCallOp]
  | | |     |                   | | |     +-[DUMMYnode]
  | | |     |                   | | |   +-[SelectOp]
  | | |     |                   | | |   | | +-[DUMMYnode]
  | | |     |                   | | |   | +-[FieldOp]
  | | |     |                   | | |   |   +-[IDNode,75,"f1"]
  | | |     |                   | | | +-[SelectOp]
  | | |     |                   | | | | | +-[DUMMYnode]
  | | |     |                   | | | | +-[FieldOp]
  | | |     |                   | | | |   +-[STNode,10,"p1"]
  | | |     |                   | | +-[VarOp]
  | | |     |                   | |   +-[STNode,18,"px"]
  | | |     |                   +-[StmtOp]
  | | |     |                     |     +-[DUMMYnode]
  | | |     |                     |   +-[VarOp]
  | | |     |                     |   | +-[STNode,17,"xx"]
  | | |     |                     | +-[AssignOp]
  | | |     |                     | | |       +-[DUMMYnode]
  | | |     |                     | | |     +-[SelectOp]
  | | |     |                     | | |     | | +-[DUMMYnode]
  | | |     |                     | | |     | +-[FieldOp]
  | | |     |                     | | |     |   +-[STNode,6,"x"]
  | | |     |                     | | |   +-[SelectOp]
  | | |     |                     | | |   | | +-[DUMMYnode]
  | | |     |                     | | |   | +-[FieldOp]
  | | |     |                     | | |   |   +-[STNode,10,"p1"]
  | | |     |                     | | | +-[VarOp]
  | | |     |                     | | | | +-[STNode,18,"px"]
  | | |     |                     | | +-[AssignOp]
  | | |     |                     | |   +-[DUMMYnode]
  | | |     |                     +-[StmtOp]
  | | |     |                       |   +-[DUMMYnode]
  | | |     |                       | +-[RoutineCallOp]
  | | |     |                       | | |   +-[DUMMYnode]
  | | |     |                       | | | +-[SelectOp]
  | | |     |                       | | | | | +-[DUMMYnode]
  | | |     |                       | | | | +-[FieldOp]
  | | |     |                       | | | |   +-[STNode,11,"f1"]
  | | |     |                       | | +-[VarOp]
  | | |     |                       | |   +-[STNode,13,"p21"]
  | | |     |                       +-[StmtOp]
  | | |     |                         |   +-[DUMMYnode]
  | | |     |                         | +-[RoutineCallOp]
  | | |     |                         | | |   +-[DUMMYnode]
  | | |     |                         | | | +-[SelectOp]
  | | |     |                         | | | | | +-[DUMMYnode]
  | | |     |                         | | | | +-[FieldOp]
  | | |     |                         | | | |   +-[STNode,7,"f0"]
  | | |     |                         | | +-[VarOp]
  | | |     |                         | |   +-[STNode,3,"Point0"]
  | | |     |                         +-[StmtOp]
  | | |     |                           +-[DUMMYnode]
  | | |   +-[BodyOp]
  | | |   | |       +-[DUMMYnode]
  | | |   | |     +-[CommaOp]
  | | |   | |     | | +-[DUMMYnode]
  | | |   | |     | +-[TypeIdOp]
  | | |   | |     |   +-[STNode,8,"Point1"]
  | | |   | |   +-[CommaOp]
  | | |   | |   | +-[STNode,18,"px"]
  | | |   | | +-[DeclOp]
  | | |   | | | +-[DUMMYnode]
  | | |   | +-[BodyOp]
  | | |   |   |       +-[NUMNode,22]
  | | |   |   |     +-[CommaOp]
  | | |   |   |     | | +-[DUMMYnode]
  | | |   |   |     | +-[TypeIdOp]
  | | |   |   |     |   +-[INTEGERTNode]
  | | |   |   |   +-[CommaOp]
  | | |   |   |   | +-[STNode,17,"xx"]
  | | |   |   | +-[DeclOp]
  | | |   |   | | +-[DUMMYnode]
  | | |   |   +-[BodyOp]
  | | |   |     +-[DUMMYnode]
  | | | +-[MethodOp]
  | | | | |   +-[DUMMYnode]
  | | | | | +-[SpecOp]
  | | | | | | +-[DUMMYnode]
  | | | | +-[HeadOp]
  | | | |   +-[STNode,16,"main"]
  | | +-[BodyOp]
  | |   |       +-[DUMMYnode]
  | |   |     +-[CommaOp]
  | |   |     | | +-[DUMMYnode]
  | |   |     | +-[TypeIdOp]
  | |   |     |   +-[INTEGERTNode]
  | |   |   +-[CommaOp]
  | |   |   | +-[STNode,15,"xy"]
  | |   | +-[DeclOp]
  | |   | | +-[DUMMYnode]
  | |   +-[BodyOp]
  | |     |       +-[DUMMYnode]
  | |     |     +-[CommaOp]
  | |     |     | | +-[DUMMYnode]
  | |     |     | +-[TypeIdOp]
  | |     |     |   +-[STNode,3,"Point0"]
  | |     |   +-[CommaOp]
  | |     |   | +-[STNode,14,"p20"]
  | |     | +-[DeclOp]
  | |     | | +-[DUMMYnode]
  | |     +-[BodyOp]
  | |       |       +-[DUMMYnode]
  | |       |     +-[CommaOp]
  | |       |     | | +-[DUMMYnode]
  | |       |     | +-[TypeIdOp]
  | |       |     |   +-[STNode,8,"Point1"]
  | |       |   +-[CommaOp]
  | |       |   | +-[STNode,13,"p21"]
  | |       | +-[DeclOp]
  | |       | | +-[DUMMYnode]
  | |       +-[BodyOp]
  | |         +-[DUMMYnode]
  +-[ClassOp]
    |   +-[STNode,8,"Point1"]
    | +-[ClassDefOp]
    | | |           +-[DUMMYnode]
    | | |         +-[CommaOp]
    | | |         | | +-[DUMMYnode]
    | | |         | +-[VarOp]
    | | |         |   +-[STNode,9,"t1"]
    | | |       +-[RoutineCallOp]
    | | |       | |   +-[DUMMYnode]
    | | |       | | +-[SelectOp]
    | | |       | | | | +-[DUMMYnode]
    | | |       | | | +-[FieldOp]
    | | |       | | |   +-[STNode,2,"println"]
    | | |       | +-[VarOp]
    | | |       |   +-[STNode,1,"system"]
    | | |     +-[StmtOp]
    | | |     | |     +-[DUMMYnode]
    | | |     | |   +-[CommaOp]
    | | |     | |   | +-[STRINGNode,89,"After t1="]
    | | |     | | +-[RoutineCallOp]
    | | |     | | | |   +-[DUMMYnode]
    | | |     | | | | +-[SelectOp]
    | | |     | | | | | | +-[DUMMYnode]
    | | |     | | | | | +-[FieldOp]
    | | |     | | | | |   +-[STNode,2,"println"]
    | | |     | | | +-[VarOp]
    | | |     | | |   +-[STNode,1,"system"]
    | | |     | +-[StmtOp]
    | | |     |   |   +-[DUMMYnode]
    | | |     |   | +-[RoutineCallOp]
    | | |     |   | | |   +-[DUMMYnode]
    | | |     |   | | | +-[SelectOp]
    | | |     |   | | | | | +-[DUMMYnode]
    | | |     |   | | | | +-[FieldOp]
    | | |     |   | | | |   +-[STNode,7,"f0"]
    | | |     |   | | +-[VarOp]
    | | |     |   | |   +-[STNode,10,"p1"]
    | | |     |   +-[StmtOp]
    | | |     |     |   +-[NUMNode,1024]
    | | |     |     | +-[AssignOp]
    | | |     |     | | |   +-[DUMMYnode]
    | | |     |     | | | +-[VarOp]
    | | |     |     | | | | +-[STNode,9,"t1"]
    | | |     |     | | +-[AssignOp]
    | | |     |     | |   +-[DUMMYnode]
    | | |     |     +-[StmtOp]
    | | |     |       |     +-[DUMMYnode]
    | | |     |       |   +-[CommaOp]
    | | |     |       |   | | +-[DUMMYnode]
    | | |     |       |   | +-[VarOp]
    | | |     |       |   |   +-[STNode,9,"t1"]
    | | |     |       | +-[RoutineCallOp]
    | | |     |       | | |   +-[DUMMYnode]
    | | |     |       | | | +-[SelectOp]
    | | |     |       | | | | | +-[DUMMYnode]
    | | |     |       | | | | +-[FieldOp]
    | | |     |       | | | |   +-[STNode,2,"println"]
    | | |     |       | | +-[VarOp]
    | | |     |       | |   +-[STNode,1,"system"]
    | | |     |       +-[StmtOp]
    | | |     |         |     +-[DUMMYnode]
    | | |     |         |   +-[CommaOp]
    | | |     |         |   | +-[STRINGNode,78,"Before t1="]
    | | |     |         | +-[RoutineCallOp]
    | | |     |         | | |   +-[DUMMYnode]
    | | |     |         | | | +-[SelectOp]
    | | |     |         | | | | | +-[DUMMYnode]
    | | |     |         | | | | +-[FieldOp]
    | | |     |         | | | |   +-[STNode,2,"println"]
    | | |     |         | | +-[VarOp]
    | | |     |         | |   +-[STNode,1,"system"]
    | | |     |         +-[StmtOp]
    | | |     |           +-[DUMMYnode]
    | | |   +-[BodyOp]
    | | |   | +-[DUMMYnode]
    | | | +-[MethodOp]
    | | | | |   +-[DUMMYnode]
    | | | | | +-[SpecOp]
    | | | | | | +-[DUMMYnode]
    | | | | +-[HeadOp]
    | | | |   +-[STNode,11,"f1"]
    | | +-[BodyOp]
    | |   |       +-[DUMMYnode]
    | |   |     +-[CommaOp]
    | |   |     | | +-[DUMMYnode]
    | |   |     | +-[TypeIdOp]
    | |   |     |   +-[STNode,3,"Point0"]
    | |   |   +-[CommaOp]
    | |   |   | +-[STNode,10,"p1"]
    | |   | +-[DeclOp]
    | |   | | +-[DUMMYnode]
    | |   +-[BodyOp]
    | |     |       +-[NUMNode,12]
    | |     |     +-[CommaOp]
    | |     |     | | +-[DUMMYnode]
    | |     |     | +-[TypeIdOp]
    | |     |     |   +-[INTEGERTNode]
    | |     |   +-[CommaOp]
    | |     |   | +-[STNode,9,"t1"]
    | |     | +-[DeclOp]
    | |     | | +-[DUMMYnode]
    | |     +-[BodyOp]
    | |       +-[DUMMYnode]
    +-[ClassOp]
      |   +-[STNode,3,"Point0"]
      | +-[ClassDefOp]
      | | |           +-[DUMMYnode]
      | | |         +-[CommaOp]
      | | |         | | +-[DUMMYnode]
      | | |         | +-[VarOp]
      | | |         |   +-[STNode,6,"x"]
      | | |       +-[RoutineCallOp]
      | | |       | |   +-[DUMMYnode]
      | | |       | | +-[SelectOp]
      | | |       | | | | +-[DUMMYnode]
      | | |       | | | +-[FieldOp]
      | | |       | | |   +-[STNode,2,"println"]
      | | |       | +-[VarOp]
      | | |       |   +-[STNode,1,"system"]
      | | |     +-[StmtOp]
      | | |     | |     +-[DUMMYnode]
      | | |     | |   +-[CommaOp]
      | | |     | |   | +-[STRINGNode,53,"After x="]
      | | |     | | +-[RoutineCallOp]
      | | |     | | | |   +-[DUMMYnode]
      | | |     | | | | +-[SelectOp]
      | | |     | | | | | | +-[DUMMYnode]
      | | |     | | | | | +-[FieldOp]
      | | |     | | | | |   +-[STNode,2,"println"]
      | | |     | | | +-[VarOp]
      | | |     | | |   +-[STNode,1,"system"]
      | | |     | +-[StmtOp]
      | | |     |   |         +-[NUMNode,1]
      | | |     |   |       +-[AddOp]
      | | |     |   |       | | +-[DUMMYnode]
      | | |     |   |       | +-[VarOp]
      | | |     |   |       |   +-[STNode,5,"y"]
      | | |     |   |     +-[AssignOp]
      | | |     |   |     | |   +-[DUMMYnode]
      | | |     |   |     | | +-[VarOp]
      | | |     |   |     | | | +-[STNode,5,"y"]
      | | |     |   |     | +-[AssignOp]
      | | |     |   |     |   +-[DUMMYnode]
      | | |     |   |   +-[StmtOp]
      | | |     |   |   | |     +-[NUMNode,1]
      | | |     |   |   | |   +-[AddOp]
      | | |     |   |   | |   | |   +-[DUMMYnode]
      | | |     |   |   | |   | | +-[SelectOp]
      | | |     |   |   | |   | | | | +-[DUMMYnode]
      | | |     |   |   | |   | | | +-[IndexOp]
      | | |     |   |   | |   | | |   | +-[NUMNode,1]
      | | |     |   |   | |   | | |   +-[SubOp]
      | | |     |   |   | |   | | |     | +-[DUMMYnode]
      | | |     |   |   | |   | | |     +-[VarOp]
      | | |     |   |   | |   | | |       +-[STNode,5,"y"]
      | | |     |   |   | |   | +-[VarOp]
      | | |     |   |   | |   |   +-[STNode,4,"parr0"]
      | | |     |   |   | | +-[AssignOp]
      | | |     |   |   | | | |     +-[DUMMYnode]
      | | |     |   |   | | | |   +-[SelectOp]
      | | |     |   |   | | | |   | | +-[DUMMYnode]
      | | |     |   |   | | | |   | +-[IndexOp]
      | | |     |   |   | | | |   |   | +-[DUMMYnode]
      | | |     |   |   | | | |   |   +-[VarOp]
      | | |     |   |   | | | |   |     +-[STNode,5,"y"]
      | | |     |   |   | | | | +-[VarOp]
      | | |     |   |   | | | | | +-[STNode,4,"parr0"]
      | | |     |   |   | | | +-[AssignOp]
      | | |     |   |   | | |   +-[DUMMYnode]
      | | |     |   |   | +-[StmtOp]
      | | |     |   |   |   +-[DUMMYnode]
      | | |     |   | +-[LoopOp]
      | | |     |   | | | +-[NUMNode,12]
      | | |     |   | | +-[LTOp]
      | | |     |   | |   | +-[DUMMYnode]
      | | |     |   | |   +-[VarOp]
      | | |     |   | |     +-[STNode,5,"y"]
      | | |     |   +-[StmtOp]
      | | |     |     |   +-[NUMNode,1]
      | | |     |     | +-[AssignOp]
      | | |     |     | | |   +-[DUMMYnode]
      | | |     |     | | | +-[VarOp]
      | | |     |     | | | | +-[STNode,5,"y"]
      | | |     |     | | +-[AssignOp]
      | | |     |     | |   +-[DUMMYnode]
      | | |     |     +-[StmtOp]
      | | |     |       |       +-[DUMMYnode]
      | | |     |       |     +-[VarOp]
      | | |     |       |     | +-[STNode,6,"x"]
      | | |     |       |   +-[MultOp]
      | | |     |       |   | | +-[DUMMYnode]
      | | |     |       |   | +-[VarOp]
      | | |     |       |   |   +-[STNode,6,"x"]
      | | |     |       | +-[AssignOp]
      | | |     |       | | |   +-[DUMMYnode]
      | | |     |       | | | +-[VarOp]
      | | |     |       | | | | +-[STNode,6,"x"]
      | | |     |       | | +-[AssignOp]
      | | |     |       | |   +-[DUMMYnode]
      | | |     |       +-[StmtOp]
      | | |     |         |     +-[DUMMYnode]
      | | |     |         |   +-[CommaOp]
      | | |     |         |   | | +-[DUMMYnode]
      | | |     |         |   | +-[VarOp]
      | | |     |         |   |   +-[STNode,6,"x"]
      | | |     |         | +-[RoutineCallOp]
      | | |     |         | | |   +-[DUMMYnode]
      | | |     |         | | | +-[SelectOp]
      | | |     |         | | | | | +-[DUMMYnode]
      | | |     |         | | | | +-[FieldOp]
      | | |     |         | | | |   +-[STNode,2,"println"]
      | | |     |         | | +-[VarOp]
      | | |     |         | |   +-[STNode,1,"system"]
      | | |     |         +-[StmtOp]
      | | |     |           |     +-[DUMMYnode]
      | | |     |           |   +-[CommaOp]
      | | |     |           |   | +-[STRINGNode,43,"Before x="]
      | | |     |           | +-[RoutineCallOp]
      | | |     |           | | |   +-[DUMMYnode]
      | | |     |           | | | +-[SelectOp]
      | | |     |           | | | | | +-[DUMMYnode]
      | | |     |           | | | | +-[FieldOp]
      | | |     |           | | | |   +-[STNode,2,"println"]
      | | |     |           | | +-[VarOp]
      | | |     |           | |   +-[STNode,1,"system"]
      | | |     |           +-[StmtOp]
      | | |     |             +-[DUMMYnode]
      | | |   +-[BodyOp]
      | | |   | +-[DUMMYnode]
      | | | +-[MethodOp]
      | | | | |   +-[DUMMYnode]
      | | | | | +-[SpecOp]
      | | | | | | +-[DUMMYnode]
      | | | | +-[HeadOp]
      | | | |   +-[STNode,7,"f0"]
      | | +-[BodyOp]
      | |   |       +-[DUMMYnode]
      | |   |     +-[CommaOp]
      | |   |     | | +-[DUMMYnode]
      | |   |     | +-[TypeIdOp]
      | |   |     |   +-[INTEGERTNode]
      | |   |   +-[CommaOp]
      | |   |   | +-[STNode,5,"y"]
      | |   | +-[DeclOp]
      | |   | | |     +-[NUMNode,5]
      | |   | | |   +-[CommaOp]
      | |   | | |   | | +-[DUMMYnode]
      | |   | | |   | +-[TypeIdOp]
      | |   | | |   |   +-[INTEGERTNode]
      | |   | | | +-[CommaOp]
      | |   | | | | +-[STNode,6,"x"]
      | |   | | +-[DeclOp]
      | |   | |   +-[DUMMYnode]
      | |   +-[BodyOp]
      | |     |         +-[INTEGERTNode]
      | |     |       +-[ArrayTypeOp]
      | |     |       | | +-[NUMNode,12]
      | |     |       | +-[BoundOp]
      | |     |       |   +-[DUMMYnode]
      | |     |     +-[CommaOp]
      | |     |     | |   +-[DUMMYnode]
      | |     |     | | +-[IndexOp]
      | |     |     | | | +-[DUMMYnode]
      | |     |     | +-[TypeIdOp]
      | |     |     |   +-[INTEGERTNode]
      | |     |   +-[CommaOp]
      | |     |   | +-[STNode,4,"parr0"]
      | |     | +-[DeclOp]
      | |     | | +-[DUMMYnode]
      | |     +-[BodyOp]
      | |       +-[DUMMYnode]
      +-[ClassOp]
        +-[DUMMYnode]
